components/suggest/src/db.rs

```diff
@@ -128,63 +128,72 @@ impl<'a> SuggestDao<'a> {
 
     /// Fetches suggestions that match the given query from the database.
     pub fn fetch_suggestions(&self, query: &SuggestionQuery) -> Result<Vec<Suggestion>> {
-        if let Some(suggestion) = self.fetch_yelp_suggestion(query)? {
-            return Ok(vec![suggestion]);
-        }
-
-        let keyword_lowercased = &query.keyword.to_lowercase();
-        let (keyword_prefix, keyword_suffix) = split_keyword(keyword_lowercased);
         let suggestions_limit = query.limit.unwrap_or(-1);
+        let suggestions =
+            query
+                .providers
+                .clone()
+                .into_iter()
+                .try_fold(vec![], |mut acc, provider| {
+                    let suggestions = match provider {
+                        SuggestionProvider::Amp => self.fetch_amp_suggestions(query),
+                        SuggestionProvider::Wikipedia => self.fetch_wikipedia_suggestions(query),
+                        SuggestionProvider::Amo => self.fetch_amo_suggestions(query),
+                        SuggestionProvider::Pocket => self.fetch_pocket_suggestions(query),
+                        SuggestionProvider::Yelp => {
+                            if let Some(yelp_suggestion) = self.fetch_yelp_suggestion(query)? {
+                                Ok(vec![yelp_suggestion])
+                            } else {
+                                Ok(vec![])
+                            }
+                        }
+                    }?;
+                    acc.extend(suggestions);
+                    Ok(acc)
+                });
+
+        suggestions.map(|mut suggestions| {
+            suggestions.sort_by(|a, b| {
+                let a_score = match a {
+                    Suggestion::Amp { score, .. }
+                    | Suggestion::Pocket { score, .. }
+                    | Suggestion::Amo { score, .. } => score,
+                    _ => &DEFAULT_SUGGESTION_SCORE,
+                };
+                let b_score = match b {
+                    Suggestion::Amp { score, .. }
+                    | Suggestion::Pocket { score, .. }
+                    | Suggestion::Amo { score, .. } => score,
+                    _ => &DEFAULT_SUGGESTION_SCORE,
+                };
+                b_score
+                    .partial_cmp(a_score)
+                    .unwrap_or(std::cmp::Ordering::Equal)
+            });
+            if suggestions.len() > suggestions_limit as usize {
+                suggestions.truncate(suggestions_limit as usize);
+            }
```

Comment:
so sorting after fetching from each provider came with a snag and seems a little messy. Since Wikipedia and Yelp doesn't have a score field in their `Suggestion` structs, I used the `DEFAULT_SUGGESTION_SCORE`.

Reply:
Oof, yes, the `f64` sure makes sorting cumbersome. I really like @ncloudioj's suggestion to implement `Ord` for `Suggestion`, though—then we can lift that complexity, and the default score for Wikipedia, out of here, and then use `suggestions.sort()`.